Make `Disabled` component truly disable child interactivity

getdave opened the pull request:
Currently the `Disabled` component is somewhat confusing and inflexible.

1. The `README` doesn't mention it only disables _form_ elements.
2. There is no way to extend the component to disable other types of focusable elements.

This PR 

* extends the component to disable all interactivity within the child components (specifically ability to tab to or `focus` elements)
* adds an `onDisable` callback which accepts the  detected `focusable` elements and allows developers to choose to disable other types of focusable elements.

This ~is intentionally implemented to be fully backwards compatible and doesn't break any existing implementations. It is purely~ may cause knock-on effects for components which wish to retain the current/older "only disable form elements" functionality. 

## How has this been tested?
Unit tests.


## Types of changes
New feature (breaking change which adds functionality).

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->

getdave committed:
Updates docs to make it clear as to the default scope of the component functionality

getdave committed:
Adds ability to config nodes and callback to handle additional requirements

Previously there was no easy way to extend the logic for which focusable elements the component disables. By default it only handles form elements.

Adding a onDisable callback allows developers to extend the defaults to handle other kinds of focusable elements such as anchors.

youknowriad commented on packages/components/src/disabled/README.md:
```diff
@@ -1,6 +1,8 @@
-Disabled is a component which disables descendant tabbable elements and prevents pointer interaction.
+Disabled is a component which disables descendant tabbable form input elements and prevents pointer interaction.
+
+Note: by default this component does not prevent tabbing to anchor elements. If this is required then you may utilise the `onDisable` prop to manually disable individual elements on a per node basis.
```
It feels like a bug to me and that we should also disable these by default unless there's something I'm missing.

getdave replied:
It's a known issue.

For example, where in [this code](https://github.com/WordPress/gutenberg/pull/17064/files#diff-beb3ac40e8f8c08b7c848a5d7c78a803R74) does it disable anchors?

youknowriad replied:
My understanding of the component was that it should disable every tabbable element. Not sure who can confirm/inform this? cc @jasmussen @gziolo

getdave replied:
Personally I feel a `Disabled` component should disable all interactions. Whether or not this is an a11y concern is not an implementation detail of this component. Rather it is a concern **at the point at which the component is _used_**.

Therefore I propose we make the `Disabled` component completely disabled all interactions within the wrapped component.

We could add a delegated event listener on `this.node` and `preventDefault()` on all event types. Combined with the existing `pointer-events` approach here this would completely disabled interaction. 

Noting that we _might_ also need to add `tabindex="-1"` to add focusable nodes to ensure that they cannot receive focus in the first place, else calling `preventDefault` might lead to strange focus context loss.

What do you think @youknowriad?

youknowriad replied:
I agree with the reasoning especially if we take a look at the places this component is used. I feel we should disable everything. It's like applying a nested `role="presentation"`

getdave replied:
Ok @youknowriad I've updated the component to disable the ability to focus or tab to elements within the child component.

getdave committed:
Disables all interactivity within the child component

As per the discussion below, we now disable all interactivity within the wrapped component. We catch all interaction events and also disable the focusability and tabability of interactive elements. Callback is retained for extensbility.

youknowriad commented on packages/components/src/disabled/index.js:
```diff
@@ -52,31 +53,54 @@ class Disabled extends Component {
+			focusable.setAttribute( 'tabindex', '-1' );
 		} );
+
+		onDisable( focusableNodes );
```
Should we remove ` onDisable` prop and update the README?

getdave replied:
✅  Done and done.

getdave committed:
Updates tests to cater for all interactive elements

getdave committed:
Remove unecessary onDisable prop

getdave committed:
Update README to describe new component functionality

getdave committed:
Adds usage note to readme re a11y

jasmussen commented:
Thanks for fixing this issue. When used for previews, the component is intended to let the preview as just that, an illustration of the final result. Only: it is markup instead of pixels, HTML instead of SVG. This is the same for blocks that are in an error state, or any other appropriate uses of the Disabled component; the content that is disabled is not to be interacted with, whether through keyboard or mouse. In that vein, it makes sense for it to be completely, not partially inert, and for the behavior to be the same between mouse and keyboard. I will defer to Riad and others on the technical implementation, but wanted to chime in with my opionion  👍 👍

youknowriad commented on packages/components/src/disabled/index.js:
```diff
@@ -52,30 +53,50 @@ class Disabled extends Component {
 			subtree: true,
 		} );
+
+		this.node.addEventListener( 'click', this.catchEvents );
```
What if the preview is displayed inside a Button. will this prevent the "click" from happening?

getdave replied:
Good catch! I've removed `stopPropagation`.

getdave committed:
Add fix for RichText component being focusable in FF via scripting

youknowriad commented on packages/components/src/disabled/index.js:
```diff
@@ -94,10 +99,12 @@ class Disabled extends Component {
+
+		// In FF the `<RichText>` node becomes focusable (presumanly via scripting).
+		// We manually disable this via tabindex also
+		const blockEditorEdgeCaseFocusable = '.block-editor-rich-text';
```
@ellatrix might know better how this works. apparently `focus.focusable.find( this.node );` is not detecting these nodes as focusable.

getdave replied:
...and [here is a video](https://cld.wthms.co/ZzZKPW) of a `BlockPreview` component that is wrapped in a `Disabled` component. You can see that first

* For some reason `button` with `disabled` and `tabindex="=1"` is managing to be focusable (what!?)
* `<div class="editor-rich-text block-editor-rich-text" />` is focusable

The edge case above seems to fix both issues, but I have no idea why it becomes focusable in the first place.

ellatrix replied:
I have no idea. It should be regarded as focusable because it has `contenteditable="true"` as an attribute.

ellatrix replied:
I tried `wp.dom.focus.focusable.find( document.querySelector( '[data-type="core/paragraph"]' ) )` and it works fine.

getdave replied:
I _think_ I’ve figured it out…

As I suspected, it’s a FF specific thing. I cannot replicate in other browsers.

Basically, according to [the spec](https://html.spec.whatwg.org/multipage/interaction.html#focusable-area) an element which is a **scrollable container** is focusable in FF (a bit like as though it had tabindex="0". 

Also note this from the spec:

> The CSS ‘overflow’ property’s ‘scroll’ value typically creates a scrollable region.

Now [see my video of me updating the overflow to visible](https://cld.wthms.co/5a1HEE) on the div.block-editor-rich-text node. 

See how after I stop it being a scrollable container it no longer receives focus? Yep!

This explains why it isn't detected by our `focusable` utility and why my hack fix (in this PR) above does work.

So basically we need to 

1. Find a way to detect scrollable containers in our `focus` utility
2. Update RichText to not create a scrollable container

getdave replied:
Here's an idea for how we could detect these focusable scrollable containers:

```
const context = // some DOM node here

[...context.querySelectorAll('*')].filter(el => {
	const hasValidOverflow = getComputedStyle(el).overflow !== "visible"; // ouch this is expensive!
	const hasScroll = el.clientHeight >= el.scrollHeight; // and this :(

	return hasValidOverflow && hasScroll;
});
```

I'd suggest we'd need to make this an opt in feature as you'd need to be very careful about the perf impact of interrogating all those DOM nodes.

ellatrix replied:
> It is the wrapper element,

Which wrapping element? Is it part of `RichText`?

> Update RichText to not create a scrollable container

Where is it creating a scrollable container?

getdave replied:
> Which wrapping element? Is it part of RichText?

Yes. It's the one I linked to above which is part of `RichText`. Here's the full link to be completely unambiguous:

https://github.com/WordPress/gutenberg/blob/d63306d5e518486f0c074a9c5122d6b8cf7c83aa/packages/block-editor/src/components/rich-text/index.js#L417

> Where is it creating a scrollable container?

If you watch the video, you'll see me tabbing through the preview elements (these are all standard `button`s) without being dragged into the DOM of the preview itself because it has been disabled (via this PR). When I get to "Portfolio" however I end up with my tab focus being dragged onto the element referenced above. I then inspect it in Firefox devtools and you will see it is marked as `scroll`, which is Firefox's way of telling you that it is a "scrollable container".

I've since worked out that the scrollable container it is infact being created by styles being applied by the `core/gallery` block. Specifically the `overflow-y: auto`. 

https://github.com/WordPress/gutenberg/blob/2f9d0e20775083b3f1f99da2bc0f9e55ae78e6e0/packages/block-library/src/gallery/editor.scss#L35

If this is changed to `overflow: hidden` then the focus bug is completed removed.

getdave replied:
Ok @youknowriad @ellatrix, I've updated `focusable.find()` to have an option to detect scrollable containers and then applied this on the `Disable` component to ensure we truely disable all focus/tab stops within the wrapped components.

This fixes the bug I was experiencing.

getdave committed:
Avoid preventing event bubble

Previously cancelling event bubbling meant that elements wrapping the disabled element wouldn’t trigger clicks. So if you had a button wrapping an element which was wrapped in disabled it might make the button seem unclickable.

getdave committed:
Update focusable util to include scrollable containers

 Firefox treats elements which create a scrollable region/container as "focusable". Detecting these is expensive, so this is an opt-in.
see https://html.spec.whatwg.org/multipage/interaction.html#focusable-area

getdave committed:
Update to detect and disable scrollable containers

Disable needs to detect *all* focusable elements. FF treats scrollable containers as focusable. Therefore we enable the option on the focusable util to detect these elements and then disable them.

youknowriad commented on packages/dom/src/focusable.js:
```diff
@@ -88,4 +93,20 @@ export function find( context ) {
+		} );
+
+		focusable = [ ...focusable, ...scrollableContainers ];
+	}
```
Are we concerned about potential performance impacts here?

getdave replied:
I'm probably being naive, but is there something about the merging of the two arrays that is specifically concerning? I assume because we could have duplicates in both arrays?

We could use https://lodash.com/docs/4.17.15#uniq to de-dupe to avoid unecessary Nodes being returned.

I'd value any additional guidance here.

Other than this, accessing the DOM layout properties is very expensive in terms of layout reflow/paint...etc. Unfortunately, there isn't a way around this that I'm aware of.

I've added a minor optimisation to avoid the call to check the `overflow` property if there is no scroll detected. This should help.

youknowriad replied:
Oh sorry I was not talking about merging arrays but more. The querySelector call and this

> Other than this, accessing the DOM layout properties is very expensive in terms of layout reflow/paint...etc. Unfortunately, there isn't a way around this that I'm aware of.

I understand that there's no way around it but how many times we're calling this function... is maybe where we can guarantee that it's not something recurrent that will harm us.

getdave replied:
> The querySelector call

I've replaced with this with `getElementsByTagName(*)` because it's a faster performing API.

> ...how many times we're calling this function...

getdave committed:
Minor perf optimisation to focusable

Bale early if there is no scroll to avoid unecessary additional call to check computed overflow property.

getdave committed:
Utilise getElementsByTagName to improve performance

getElementsByTagName is known to be faster than querySelectorAll. Therefore using it here will help get us a collection faster.

youknowriad commented on packages/components/src/disabled/index.js:
```diff
@@ -52,30 +53,56 @@ class Disabled extends Component {
 		} );
+
+		document.addEventListener( 'click', this.catchEvents );
+		document.addEventListener( 'focus', this.catchEvents, true );
```
Should we remove these handlers?

getdave commented on packages/components/src/disabled/index.js:
```diff
@@ -52,30 +53,56 @@ class Disabled extends Component {
 		this.node = node;
 	}
 
 	disable() {
```
Looks like `this.disable` is called repeatedly via the `MutationObserver` because each time we modify a node it triggers a new mutation thereby re-triggering `this.disable`. It is going crazy!

I tried cancelling and requeueing the observer at the start of `this.disable` and then restarting it at the end but that didn't seem to do anything unless I added a timeout of 2000ms before re-adding the observer.

What I need is a local state variable to block running `this.disable` if it is already running.

getdave replied:
> What I need is a local state variable to block running this.disable if it is already running.

Tried this and it didn't work.

jorgefilipecosta replied:
> Looks like this.disable is called repeatedly via the MutationObserver because each time we modify a node it triggers a new mutation thereby re-triggering this.disable. It is going crazy!

Watching the code it looks like this problem is unexpected as it was handled by the usage of debounce:
```
// Debounce re-disable since disabling process itself will incur
// additional mutations which should be ignored.
```
Does adding a timeout to debounce fixes the problem?

If not, I guess we can try something safer but a little bit complex. The code that applies mutations is part of the mutation handler event what if on the mutation handler event we disable the mutation observer. We schedule the execution of the disable() operation and the disable operation after apply the mutations enables the mutation observer.
If the last suggestion also does not work I guess we can try a three scheduling approach: the mutation handler schedules a function "A". Function "A" disables the mutation observer and schedules function "B". Function "B" applies the mutations to the dom and schedules function "C". Function "C" reenables the mutation observer.

The suggestions seem hacky but if one of them works we may then understand better what's happening and may be able to find a better solution.

getdave replied:
I've tried some of these previously, but let me come at it again with a clear head and see how I get on. Thanks @jorgefilipecosta 👍

getdave replied:
Going to be looking into this soon.

getdave commented:
Closing this one as stale.
